### PR TITLE
Refactor Server Connector to make it able to interact with external extensions #370

### PR DESCRIPTION
**Pull request by lstocchi**
#370 

This is the first prototype to make server connector works with external rsp providers. 
To make some tests you can use this external rsp provider https://github.com/lstocchi/external-rsp-provider-sample 
I also created this npm package which is useful to create a new rsp provider https://github.com/lstocchi/vscode-server-connector-api

Currently the external rsp provider can only start a server. Tests are failing inside server connector, name of external rsp provider is hard-coded and i need to write a lot of tests for the new stuff. I'm going to fix them in the next days but i'm creating this PR so you can start checking the code.

### CODE REVIEW

**Comment by lstocchi** [comment]
> `serverExplorer.RSPServersStatus`: why use a capital letter to start an instance var?

because it's an acronym. 

Microsoft suggests to use capital letters only if acronyms consist of 2 chars. In this case they are 3 so following their rule it should be RspServersStatus - https://stackoverflow.com/a/15526526 - but i found that in some occasion they don't follow this rule at all ... e.g CloudExplorerAPI, ClusterProviderAPI (taken from kubernetes)

From an aesthetic point of view i prefer all capital letters but we should definetly stick with a coding style. If you have some preference please let me know.

File: src/api/implementation/rspProviderAPI.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    public async deregisterRSPProvider(id: string): Promise<void> {
+        if (!id) {
+            const error = 'Unable to remove RSP provider - Id is not valid.';
+            vscode.window.showErrorMessage(error);
+            return Promise.reject(error);
+        }
+        const serversExplorer = ServerExplorer.getInstance();
+        serversExplorer.RSPServersStatus.get(id).rspserverstdout.dispose();
```
Comment (robstryker) [inline src/api/implementation/rspProviderAPI.ts]: Any chance for NPEs here or in next line?

File: src/extension.ts [hunk 2]
```diff
@@ -105,39 +69,27 @@ function registerCommands(commandHandler: CommandHandler, context: vscode.Extens
...
 
-function stopServer(val: Protocol.ServerState) {
-    const oneStat: Protocol.ServerState = val;
+function stopServer(val: ServerStateNode) {
+    const oneStat: ServerStateNode = val;
     const stateNum = oneStat.state;
     if (stateNum !== ServerState.UNKNOWN
       && stateNum !== ServerState.STOPPED
       && stateNum !== ServerState.STOPPING) {
+        const client: RSPClient = serversExplorer.RSPServersStatus.get(oneStat.server.id).client;
```
Comment (robstryker) [inline src/extension.ts]: check for NPEs here and next line?

**Comment by lstocchi** [comment]
Finished writing additional tests. Now the code coverage is over 55% and we still need to add the other PR which contains tests related to the editor.

**Comment by lstocchi** [comment]
Update: thanks to @robstryker who uploaded two different rsp server (one for wildfly/eap, one for minishift/cdk) i was able to create a new rsp-provider

1) https://github.com/lstocchi/minishift-rsp-provider-sample
2) https://github.com/lstocchi/wildfly-eap-rsp-provider-sample

Done some quick test and Server Connector UI seems working great with multiple rsp providers. 

image

If you want to do some tests, just download the rsp-provider samples, create a vsix packages from them (cmd -> vsce package) and upload them as additional extensions.

**Comment by lstocchi** [comment]
Closing ... PR can be found on new repo https://github.com/redhat-developer/vscode-rsp-ui

**Pull request closed without merging**